### PR TITLE
Add JoinableFactory interface and use it in the query stack.

### PR DESCRIPTION
**Pull request by gianm**
Also includes InlineJoinableFactory, which enables joining against
inline datasources. This is the first patch where a basic join query
actually works. It includes integration tests.

### CODE REVIEW

File: integration-tests/docker/broker.conf [hunk 1]
```diff
@@ -11,6 +11,7 @@ command=java
   -Ddruid.host=%(ENV_HOST_IP)s
   -Ddruid.zk.service.host=druid-zookeeper-kafka
   -Ddruid.processing.buffer.sizeBytes=25000000
+  -Ddruid.query.groupBy.maxOnDiskStorage=300000000
```
Comment (suneet-s) [inline integration-tests/docker/broker.conf]: Why was this change needed?

File: processing/src/main/java/org/apache/druid/segment/ReferenceCounter.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package org.apache.druid.segment;
+
+import java.io.Closeable;
+
+/**
+ * An interface to reference-counted objects. Used by {@link ReferenceCountingSegment}.
+ */
+public interface ReferenceCounter
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/segment/ReferenceCounter.java]: Does this need to be thread safe?

File: processing/src/main/java/org/apache/druid/segment/ReferenceCountingSegment.java [hunk 3]
```diff
@@ -167,6 +166,30 @@ public void close()
...
+      {
+        return ReferenceCountingSegment.this.decrementOnceCloseable();
+      }
+
+      @Override
+      public void decrement()
+      {
+        ReferenceCountingSegment.this.decrement();
+      }
+    };
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/segment/ReferenceCountingSegment.java]: Make `ReferenceCountingSegment` implement `ReferenceCounter`. Then this function can just be `return this;`

Which also means less inner classes and less lines of code to read in a delta 🙃

File: processing/src/main/java/org/apache/druid/segment/join/HashJoinSegment.java [hunk 4]
```diff
@@ -50,15 +49,9 @@ public HashJoinSegment(
     this.baseSegment = baseSegment;
     this.clauses = clauses;
 
-    // Verify no clauses would shadow the special __time field.
-    for (JoinableClause clause : clauses) {
-      if (clause.includesColumn(ColumnHolder.TIME_COLUMN_NAME)) {
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/segment/join/HashJoinSegment.java]: Does this mean we now allow joins on the time column?

File: processing/src/main/java/org/apache/druid/query/ReferenceCountingSegmentQueryRunner.java [hunk 5]
```diff
@@ -22,39 +22,41 @@
...
-  private final ReferenceCountingSegment adapter;
+  private final Segment segment;
+  private final ReferenceCounter segmentReferenceCounter;
   private final SegmentDescriptor descriptor;
 
   public ReferenceCountingSegmentQueryRunner(
       QueryRunnerFactory<T, Query<T>> factory,
-      ReferenceCountingSegment adapter,
+      Segment segment,
+      ReferenceCounter segmentReferenceCounter,
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/ReferenceCountingSegmentQueryRunner.java]: note to self: Do these 2 need to be tightly coupled? DIg in a little more here

File: processing/src/main/java/org/apache/druid/segment/join/Joinables.java [hunk 6]
```diff
@@ -52,4 +61,56 @@ public static boolean isPrefixedBy(final String columnName, final String prefix)
...
+  /**
+   * Creates a Function that maps base segments to {@link HashJoinSegment} if needed (i.e. if the number of join
+   * clauses is > 0). If mapping is not needed, this method will return {@link Function#identity()}.
+   *
+   * @param clauses            pre-joinable clauses
+   * @param joinableFactory    factory for joinables
+   * @param cpuTimeAccumulator an accumulator that we will add CPU nanos to; this is part of the function to encourage
+   *                           callers to remember to track metrics on CPU time required for creation of Joinables
+   */
+  public static Function<Segment, Segment> createSegmentMapFn(
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/segment/join/Joinables.java]: missing unit tests?

File: core/src/main/java/org/apache/druid/utils/JvmUtils.java [hunk 7]
```diff
@@ -80,13 +82,25 @@ public static long safeGetThreadCpuTime()
...
    * @throws UnsupportedOperationException if the Java virtual machine does not support CPU time measurement for
-   * the current thread.
+   *                                       the current thread.
    */
   public static long getCurrentThreadCpuTime()
   {
     return THREAD_MX_BEAN.getCurrentThreadCpuTime();
   }
 
+  public static <T> T safeAccumulateThreadCpuTime(final AtomicLong accumulator, final Supplier<T> function)
```
[the remark this answers is not in the document]
Comment (gianm) [reply core/src/main/java/org/apache/druid/utils/JvmUtils.java]: 👍

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTask.java [hunk 8]
```diff
@@ -206,6 +206,7 @@ public Appenderator newAppenderator(FireDepartmentMetrics metrics, TaskToolbox t
         toolbox.getSegmentAnnouncer(),
         toolbox.getEmitter(),
         toolbox.getQueryExecutorService(),
+        toolbox.getJoinableFactory(),
```
[the remark this answers is not in the document]
Comment (gianm) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTask.java]: Yes, GitHub issues are a good place to put stuff like this. I think there's even a 'refactoring' label so the ideas can be searched.

File: indexing-service/src/test/java/org/apache/druid/indexing/common/task/TestAppenderatorsManager.java [hunk 9]
```diff
@@ -78,6 +81,7 @@ public Appenderator createRealtimeAppenderatorForTask(
         segmentAnnouncer,
         emitter,
         queryExecutorService,
+        NoopJoinableFactory.INSTANCE,
```
[the remark this answers is not in the document]
Comment (gianm) [reply indexing-service/src/test/java/org/apache/druid/indexing/common/task/TestAppenderatorsManager.java]: Oops, yes, you're right.

File: integration-tests/docker/broker.conf (hunk 1 above)
Comment (gianm) [reply integration-tests/docker/broker.conf]: Because the new integration tests I added needed more aggregation space than the 25MB of memory allocated by the previous line. I didn't want to allocate more memory (and therefore raise the memory requirements of integration tests) so instead I am configuring it to use disk for the spillover.

File: processing/src/main/java/org/apache/druid/segment/ReferenceCounter.java (hunk 2 above)
Comment (gianm) [reply processing/src/main/java/org/apache/druid/segment/ReferenceCounter.java]: Yes, it does. I'll add some comments for it.

File: processing/src/main/java/org/apache/druid/segment/ReferenceCountingSegment.java (hunk 3 above)
Comment (gianm) [reply processing/src/main/java/org/apache/druid/segment/ReferenceCountingSegment.java]: Oh, great idea!

File: processing/src/main/java/org/apache/druid/segment/join/HashJoinSegment.java (hunk 4 above)
Comment (gianm) [reply processing/src/main/java/org/apache/druid/segment/join/HashJoinSegment.java]: Yes, we do, although that's not what this means. This same check is still being applied, but in the constructor of JoinableClause (it calls `Joinables.validatePrefix`).

File: processing/src/main/java/org/apache/druid/segment/join/Joinables.java (hunk 6 above)
Comment (gianm) [reply processing/src/main/java/org/apache/druid/segment/join/Joinables.java]: I'll add some. I'm thinking of adding tests that are a bit on the basic side, because full-on tests require a lot of scaffolding similar to HashJoinSegmentTest. I'm not sure if adding that here adds much value over the tests that are already in HashJoinSegmentTest. So I'm planning to focus on testing the functionality that HashJoinSegmentTest _doesn't_ test (error conditions, special cases).

Let me know what you think. If you think I should add more tests beyond the ones I'm about to add, please let me know.

**Comment by gianm** [comment]
@suneet-s I've pushed up a commit addressing your comments; let me know what you think.

**Comment by gianm** [comment]
Thanks for review folks
